remove unnecessary clone() when .grad is None

FDecaYed opened the pull request:
Currently gradient is copied into .grad if it is None. This PR aim to remove the copy when it is not absolutely needed.

It is generally an improvement of speed and memory usage. And here is a case it may help a lot:
Normally, people do optimizer.zero_grad() every minibatch before backward. It will translate into a memset, and later a point-wise add.
When there is some large weight in the network, one optimization people can always do is set parameter.grad to None instead of zero_grad. This will remove memset and change point-wise add to a memcpy.
Here is result running following script on V100 GPU. It is 100 iterations of forward/backward/zero_grad on single 1-billion word benchmark size embedding.
`Zero grad: 2.123847723007202`
`None grad: 1.3342866897583008`

With the backend change of this PR, the unnecessary memcpy is removed, thus further speed up is achieved.
`Zero grad: 2.124978542327881`
`None grad: 0.4396955966949463`

[benchmark.txt](https://github.com/pytorch/pytorch/files/2341800/benchmark.txt)

Some details on the code change:
.detach() is used because we need to get rid of new_grad being a view without copy data. This should be safe in first-order only mode.
data need to be contiguous, otherwise `grad_variable.data() += new_grad.data();` below will fail.
Only the last variable that has reference to the temp gradient will grab its buffer.

@ngimel, @mcarilli  and @mruberry helped on finalizing this PR.

FDecaYed committed:
remove unnecessary clone() when .grad is None

SsnL commented:
Nice work. But I want to comment on this: 
> When there is some large weight in the network, one optimization people can always do is set parameter.grad to None instead of zero_grad. 

Actually you probably don't want to do that, at least not in current code base. Our optimization code assumes that parameters with .grad being `None` are new parameters, and not only skips gradient steps, but also skips things like weight decay, running stats estimation as well. So you are getting incorrect behavior here.

ngimel commented:
This should not be an issue if all the model parameters are updated at every iteration, right? Because then the optimizer will never see a `None` gradient, it will be assigned to a correct value during backprop.

SsnL commented:
@ngimel Yes, you are right. :)

ezyang commented:
Nice work. Is there an easy way we can test that this optimization is actually firing? For example, stash the memory address of the grad before and after accumulation, and check that it was actually moved over to grad without doing a memcpy.

mcarilli commented:
@ezyang yes, imo the easiest way to do it is to register a hook on the variable that does nothing but print the data_ptr() of its argument (and doesnt return anything). Later, after backward is done, compare this printed data_ptr to variable.grad.data_ptr().  If the optimization worked, the data_ptrs should match.

Of course if the variable also has other hooks registered on it, which may modify the gradient out of place and return something new, this is no longer guaranteed, but for a minimal working test, there’s no reason to register other hooks.

SsnL commented:
Another way to test would be to write a custom autograd Function, and compare the data_ptr of the returned grad, and the input.grad.

FDecaYed commented:
I'm working on add some test for this functionality. Where should I put the test?

ngimel commented:
test_autograd.py?

SsnL commented:
yes test_autograd.py

FDecaYed committed:
add test for remove grad copy

FDecaYed commented:
I have added test_no_grad_copy to test_autograd.py, which tests:
- non-contiguous gradient is handled correctly
- manually setting .grad to None works
- produced .grad with same result are using different buffer
- one of produced .grad did not do memcpy when no copy conditions are met

apaszke commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -40,7 +40,20 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
+      // then check not sparse before is_contiguous
+      // then check contiguous, otherwise later in place accumulation may fail
+      // and lastly, check it is the last reference before we grab it
+      // we created new_grad above so ref count should be 2
```
Can we move the variable into `new_grad`, or make it reference, such that we can assert that the use count is exactly 1?

FDecaYed replied:
It seems to me we can make new_grad reference and assert use_count 1. It is actually better since this way we will also correctly catch new_grad that produced by hook with use_count==1.
Trying to modify grad in-place with hook will modify all grads sharing the buffer, what's your opinion on that? This behavior should not be changed by make new_grad reference though. So it is probably an unrelated question.

apaszke replied:
Hooks should generally never modify their arguments, but we're not checking for that (doing so is UB)

apaszke commented on test/test_autograd.py:
```diff
@@ -2473,6 +2473,38 @@ def test_symeig_no_eigenvectors(self):
+        a = torch.randn(5, 6, requires_grad=True)
+        b = torch.randn(5, 6, requires_grad=True)
+        # sum should not trigger no copy since it produce non-contiguous grad
+        MyFunc.apply(a, b).sum().backward()
```
You're making that assumption based on the particular implementation of grad of sum. It would be much nicer to implement a function that returns a non-contiguous grad itself.

SsnL replied:
Or just do `.backward(my_non_contig_grad)`.

FDecaYed replied:
@SsnL It occurs to me that passing in grad to backward manually will change its use_count when we decide whether to clone it, so we are not testing contiguous. So I will go with first approach now.

apaszke commented on test/test_autograd.py:
```diff
@@ -2473,6 +2473,38 @@ def test_symeig_no_eigenvectors(self):
+        MyFunc.apply(a, b).sum().backward()
+        self.assertFalse(a.grad.data_ptr() == MyFunc.static_grad_ptr)
+        self.assertFalse(b.grad.data_ptr() == MyFunc.static_grad_ptr)
+        # manually setting .grad to None
```
nit: that's pretty clear from the next line 😄

FDecaYed committed:
address review comments:
change new_grad to reference and check for use_count==1
change test depend on sum() implementation

apaszke commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -33,14 +33,26 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     return {};
 
-  auto new_grad = grads[0];
+  auto& new_grad = grads[0];
```
I don't think this is correct anymore. If there are variable hooks, this will get assigned a reference to the memory of a temporary vector that the hook returned, and will outlive it. That's undefined behavior. We should really just `std::move(grads[0])` here

FDecaYed replied:
std::move should work but I think reference works too.since you cannot reassign reference in c++, the hook line is basically copying hook_returned_vector[0] into grads[0] and new_grad is still reference to grads[0].

apaszke reviewed (dismissed):
LGTM, but introduces dangling references that need to be fixed

FDecaYed committed:
change to std::move(grads[0])
